wiring/inc/spark_wiring_ble.h

```diff
@@ -0,0 +1,460 @@
+    size_t serviceUUID(BleUuid* uuids, size_t count) const;
+    size_t customData(uint8_t* buf, size_t len) const;
+
+    size_t operator () (uint8_t* buf, size_t len) const {
+        return get(buf, len);
+    }
+
+    bool contains (uint8_t type) const;
+
+private:
+    uint8_t selfData_[BLE_MAX_ADV_DATA_LEN];
+    size_t selfLen_;
+
+    size_t serviceUUID(uint8_t type, BleUuid* uuids, size_t count) const;
+    static size_t locate(const uint8_t* buf, size_t len, uint8_t type, size_t* offset);
+};
+
+
+class BleCharacteristic {
+public:
+    BleCharacteristic();
+    BleCharacteristic(const BleCharacteristic& characteristic);
+    BleCharacteristic(const char* desc, BleCharacteristicProperties properties, OnDataReceivedCallback callback=nullptr);
+
+    template<typename T>
+    BleCharacteristic(const char* desc, BleCharacteristicProperties properties, T charUuid, T svcUuid, OnDataReceivedCallback callback=nullptr) {
+        BleUuid cUuid(charUuid);
+        BleUuid sUuid(svcUuid);
+        construct(desc, properties, cUuid, sUuid, callback);
+    }
+
+    ~BleCharacteristic();
+
+    BleCharacteristic& operator = (const BleCharacteristic& characteristic);
+
+    bool valid(void) {
+        return impl() != nullptr;
+    }
+
+    BleUuid UUID(void) const;
+    BleCharacteristicProperties properties(void) const;
+
+    size_t getValue(uint8_t* buf, size_t len) const;
+    size_t getValue(String& str) const;
+
+    template<typename T>
+    size_t getValue(T* val) const {
+        size_t len = sizeof(T);
+        return getValue(reinterpret_cast<uint8_t*>(val), len);
+    }
+
+    size_t setValue(const uint8_t* buf, size_t len);
+    size_t setValue(const String& str);
+    size_t setValue(const char* str);
+
+    template<typename T>
+    size_t setValue(T val) {
+        uint8_t buf[BLE_MAX_ATTR_VALUE_PACKET_SIZE];
+        size_t len = sizeof(T) > BLE_MAX_ATTR_VALUE_PACKET_SIZE ? BLE_MAX_ATTR_VALUE_PACKET_SIZE : sizeof(T);
+        for (size_t i = 0, j = len - 1; i < len; i++, j--) {
+            buf[i] = reinterpret_cast<const uint8_t*>(&val)[j];
+        }
+        return setValue(buf, len);
+    }
+
+    void onDataReceived(OnDataReceivedCallback callback);
+
+    BleCharacteristicImpl* impl(void) const {
+        return impl_.get();
+    }
+
+private:
+    std::shared_ptr<BleCharacteristicImpl> impl_;
+    void construct(const char* desc, BleCharacteristicProperties properties, BleUuid& charUuid, BleUuid& svcUuid, OnDataReceivedCallback callback);
+};
+
+
+class BleService {
+public:
+    BleService();
+    BleService(const BleUuid& uuid);
+    ~BleService();
+
+    BleServiceImpl* impl(void) const {
+        return impl_.get();
+    }
+
+private:
+    std::shared_ptr<BleServiceImpl> impl_;
+};
+
+
+class BleScanResult {
+public:
+    BleAddress address;
+    BleAdvertisingData advertisingData;
+    BleAdvertisingData scanResponse;
+    int8_t rssi;
+};
+
+
+class BlePeerDevice {
+public:
+    BleDeviceRoles role;
```

Comment:
Should these be public? We usually provide accessor methods.

Reply:
These members will be set internally, using setters and getters would be a little overhead.